Cover refactoring

rospogrigio opened the pull request:
Inspired from #10.
Positioning option has 3 types of values: None (no positioning), Yes (standard HA positioning) and Fake (time-based, internally handled).

rospogrigio commented:
@postlund I need some help with this: when using config flow to configure a cover, I am now getting a "user input malformed" error at the top, but I can't really figure out what is wrong with my setup... can you please give me a push?

Also @andrewmeepos, can you please test this with your device and give a feedback? Only YAML works at the moment, and please notice that the structure has totally changed, refer to what is written at the top of `__init.py__` .

andrewmeepos commented:
@rospogrigio 

First, the good news. Setting `positioning_mode: "yes"` reports the current position correctly, and lets me set a new position correctly.

However, there are two substantial issues:

1. Setup via YAML fails if the numerical configs (`currpos_dps`, `setpos_dps`, `id`) are not inputted as numbers (e.g., without parenthesis around them) and the string configs (`positioning_mode`, `open_cmd`, `close_cmd`, `stop_cmd`) are not inputted as strings (e.g., with parenthesis around them). The config inputs should be accepted however they are inputted, whether with or without parenthesis, and localtuya should sanitize the data to convert the config values into the appropriate format.

2. The "stop" button sends a "close" command. 

Here is my current config.yaml:

```localtuya:
  - host: 192.168.1.29
    device_id: REDACTED
    local_key: REDACTED
    friendly_name: localtuyatest
    protocol_version: "3.3"
    entities:
      - platform: cover
        friendly_name: covertest
        id: 1
        open_cmd: "open"
        close_cmd: "close"
        stop_cmd: "close"
        positioning_mode: "yes"
        currpos_dps: 3
        setpos_dps: 2
```

Let's start with a scenario where my blind is fully closed (position is 100). If I hit the open button from lovelace's UI, the following correct command is sent:
```
2020-09-30 21:40:35 DEBUG (SyncWorker_48) [custom_components.localtuya.cover] Launching command open to cover 
2020-09-30 21:40:35 DEBUG (SyncWorker_48) [custom_components.localtuya.pytuya] Sending command set (device type: type_0a)
2020-09-30 21:40:35 DEBUG (SyncWorker_48) [custom_components.localtuya.pytuya] paylod=b'{"devId":"REDACTED","uid":"REDACTED","t":"1601527235","dps":{"1":"open"}}'
```

However, as it's opening, if I then hit the stop button from lovelace's UI, the following incorrect command is sent to the device. For some reason, it's sending `close` instead of `stop`.
```
2020-09-30 21:40:40 DEBUG (SyncWorker_30) [custom_components.localtuya.cover] Launching command close to cover 
2020-09-30 21:40:40 DEBUG (SyncWorker_30) [custom_components.localtuya.pytuya] Sending command set (device type: type_0a)
2020-09-30 21:40:40 DEBUG (SyncWorker_30) [custom_components.localtuya.pytuya] paylod=b'{"devId":"REDACTED","uid":"REDACTED","t":"1601527240","dps":{"1":"close"}}'
```

postlund commented:
@rospogrigio Hmm, I'm not sure I agree with you. The YAML config should be validated as much as possible with voluptuous to remove the need for manual verification in the code. That is the Home Assistant way and has always been. Since a DP index is a number, it should be validated as a number as well. The other keys will be validated according to whatever is defined in `flow_schema`, which I also consider to be correct.

Why did you remove `stop_cmd` btw? Isn't it better to keep it configurable in case someone needs to change it?

postlund commented on custom_components/localtuya/config_flow.py:
```diff
@@ -39,9 +39,9 @@
+        vol.Required(CONF_LOCAL_KEY): str,
         vol.Required(CONF_HOST): str,
         vol.Required(CONF_DEVICE_ID): str,
-        vol.Required(CONF_LOCAL_KEY): str,
```
Doesn't matter that much, but was this necessary?

rospogrigio replied:
@postlund Well, it's just a matter of speed if you need to configure many devices (and of my eternal laziness 😄 ), since the only param to be input after the name is the key, it just saves a couple of "tabs" during configuration. I am lazy, I now 😆

postlund replied:
Yeah, ok, it's not very pretty in the UI as it appears out-of-context 😉

rospogrigio replied:
I partially agree, actually in my mind the first two params should be the only changeable (name and key) and the others should be disabled, so it's logically divided into two blocks of params...

postlund commented on custom_components/localtuya/const.py:
```diff
@@ -17,6 +17,13 @@
+CONF_POSITIONING_MODE = "positioning_mode"
+CONF_CURRPOS = "currpos_dps"
+CONF_SETPOS = "setpos_dps"
+CONF_MODE_NONE = "none"
```
These `CONF_MODE_` constants are not configuration options (I.e. they are not stored as "keys" in the configuration) and shall thus not be named differently. It's also not obvious where they belong. I suggest that you change it to `COVER_MODE_` instead and move them to `cover.py` as they aren't used anywhere else.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -26,14 +34,23 @@
 DEFAULT_OPEN_CMD = "on"
 DEFAULT_CLOSE_CMD = "off"
 DEFAULT_STOP_CMD = "stop"
```
You did remove the stop command, so you should remove this too in that case. I think `stop_cmd` is in the example YAML in `__init__.py` as well, which you can remove.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -26,14 +34,23 @@
-        vol.Optional(CONF_STOP_CMD, default=DEFAULT_STOP_CMD): str,
+        vol.Optional(CONF_OPEN_CMD, default=DEFAULT_OPEN_CMD): vol.In(["on", "open"]),
+        vol.Optional(CONF_CLOSE_CMD, default=DEFAULT_CLOSE_CMD): vol.In(
+            ["off", "close"]
```
We can do this of course, it's good to validate properly. But do we know there are no other commands? We can of course wait for issues about it and add as needed.

rospogrigio replied:
I don't know and I can't imagine, but I believe it is more safe (and comfortable for the configuration) to have these as preset instead of needing to type them.

postlund replied:
If we know the values already, that is better of course. Otherwise we can leave the margin of error on the user with the benefit of flexibility (by allowing custom values). But it's probably just a handful of values.

rospogrigio replied:
Yep, and for the sake of speed (in case someone has to configure dozens of devices, here comes my laziness again 🤣 ) I'd be considering to have a single configuration for "on/off" vs. "open/close" couples of commands since I believe they are linked together... agree?

postlund replied:
Hmm, I'm not sure I understand what you mean? 😎

rospogrigio replied:
Merge the config options into a thing like "CONF_OPENCLOSE_CMD" with allowed values "on_off" / "open_close", and adapt the code coherently. Just one setting instead of two (and the lazy part of me celebrates 😆 ).

postlund replied:
A better way I would say would be to suggest defaults (I've been thinking about it before, but not prioritized it). Since we generate the input fields/schema in `flow_schema`, we can also provide whatever defaults we want. All the datapoints and values are passed in, so we can just scan those values and see if we find "on" or "open" and pre-fill whichever we find. It's unlikely we have overlapping datapoints at the same time.

rospogrigio replied:
Mmm, this time it is I to be not sure to have understood what you mean... 🙄

postlund replied:
The variable `dps` that is passed in to `flow_schema` is a list of strings with all supported datapoints for the device. Lets assume DP 6 is used to indicate open/on, you would find `6 (value: open)` as an example in there. We would then iterate the list (for each config option) and look for any of known values. So, for `CONF_OPEN_CMD` we would look at all datapoint, searching for `value: open` and `value: on`. Since we will find `value: open` according to example above, we can use that as a default value. It's not perfect but probably work good enough.

rospogrigio replied:
Sorry but I don't get this... on/open are commands that are supposed to be sent to the dps, not read from a dps (i.e., there si no DPS 6 with "open" value). So, unless the over is already opening/closing during the configuration, you would not read the info anywhere. Am I understanding badly?

postlund replied:
Ah, right, my bad. This only works for "state" variables. Is the state of the cover just a bool then? Or could we use that to determine what to use?

rospogrigio replied:
No, it returns "stop", "on/open" or "off/close". But when it is not operating it is "stop", of course. So in order to detect it you should go to the device and manually press the open or close buttons (assuming there is a manual control, of course). Not very comfortable...

postlund replied:
Nah, that's not good enough. Too much work to implement that, not worth it.

rospogrigio commented:
I removed the "stop_cmd" configuration since in the current use cases the same command is needed ("stop"). I would re-introduce it in case it is actually needed (i.e., we find some actual device that needs a different command, which I find it hard to imagine at the moment).

Regarding @andrewmeepos issue, I found out that the value "yes" seems to be some kind of keyword that considers it to be a boolean and then for some reason it breaks the schema validation of the other parameters. I changed it to "position" and now everything seems to work for me, please confirm on your side.

Still, I get the "user input malformed" and I can't figure out what is going wrong with the config flow, since it seems very similar to what is done for the switches... @postlund do you have any idea of what can be wrong?
Thank you

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -71,7 +88,8 @@ def __init__(
         super().__init__(device, config_entry, switchid, **kwargs)
         self._state = None
-        self._position = 50
+        self._current_cover_position = 50
```
We have a few of these constants in various platforms with hardcoded values. I believe we should just set them to `None` since we don't know the value yet and wait for an update. It's generally ok to return `None` the the properties to signal that a value is not available.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -81,72 +99,68 @@ def __init__(
+        """Return if the cover is open or not."""
+        if self._config[CONF_POSITIONING_MODE] != CONF_MODE_POSITION:
+            return None
+        else:
```
You don't need this else here due to return. There's usually a lint warning about this...

rospogrigio replied:
> Ok, so, the schema validation problem happens because `float` validates [0, 1[ for some reason. 

Gosh, I could have never imagined that. How did you figure this out, for curiosity??? The check_config script does not give validation errors...

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -81,72 +99,68 @@ def __init__(
-            return True
-        return None
+        if self._config[CONF_POSITIONING_MODE] != CONF_MODE_POSITION:
+            return None
```
Same.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -81,72 +99,68 @@ def __init__(
+
+        elif self._config[CONF_POSITIONING_MODE] == CONF_MODE_POSITION:
+            converted_position = int(kwargs[ATTR_POSITION])
+            if converted_position in range(0, 101) and self.has_config(CONF_SETPOS):
```
Use `0 < converted_position <= 100` instead, seems excessive to look for an integer in a list (even if it's a generator in this case).

rospogrigio replied:
Oooh, you are super right, I just copied it from the original PR and my knowledge of python was not enough to notice that this equals to a search in a list... 😮

postlund commented:
@rospogrigio Yeah, I can get behind that. Let's add it again when we need it.

That's part of YAML. It has a few builtin converters, like on/off will be converted to true/false and 0x123 will be converted to decimal automatically. If you want something to certainly be kept as a string, you need to quote it. I've been bitten by this before as well (because of the last one) but it's not something we should sacrifice validation for to workaround. We need to be clear about it in the instructions.

I will have a look at the validation.

postlund commented:
Ok, so, the schema validation problem happens because `float` validates [0, 1[ for some reason. You have to provide a range (which in itself is a good idea anyway) as a hack, like I did in `sensor`:

https://github.com/rospogrigio/localtuya-homeassistant/blob/2e46d65f9a89496bcfe512d462dc83ad5aaedba1/custom_components/localtuya/sensor.py#L28

rospogrigio commented:
> Ok, so, the schema validation problem happens because `float` validates [0, 1[ for some reason. 

Gosh, I could have never imagined that. How did you figure this out, for curiosity??? The check_config script does not give validation errors...

postlund commented:
> > Ok, so, the schema validation problem happens because `float` validates [0, 1[ for some reason.
> 
> Gosh, I could have never imagined that. How did you figure this out, for curiosity??? The check_config script does not give validation errors...

Looking at other integrations in the Home Assistant repo and asking on discord.

rospogrigio commented:
> Looking at other integrations in the Home Assistant repo and asking on discord.

Well, then I think somebody has to work on detailing error reporting... 😉

postlund commented:
> > Looking at other integrations in the Home Assistant repo and asking on discord.
> 
> Well, then I think somebody has to work on detailing error reporting... 😉

It would be great if the data entry flow could tell you what is wrong... You can provide error messages for individual fields if you validate manually, but things automatically validated by voluptuous you get that standard "malformed" error.

rospogrigio commented:
OK I fixed all your comments and unified open_close commands, please @andrewmeepos confirm everything is working on your side (please also try config flow!) and @postlund review.

Also, a couple of questions outside of this matter for you @postlund:
1) weeks ago I submitted a logo for localtuya custom integration as you suggested but it was never approved... is there somebody I should ask?
2) I tried the "Reload" button on Integrations created via YAML import, after changing the YAML file of course... but it doesn't work (it just receives the same data as before the YAML file was saved): is it supposed to be working? Is it a bug or is it still not a feature? How can I fix it?
Thank you guys, I really love how this is evolving 😉

postlund commented:
@rospogrigio I will have a look later today, when I have some time!

I don't know who is responsible for that repo. Try looking at some old, closed PRs and figure out who usually merge them. Maybe you can ping someone.

Reloading requires the `reload` service to work. I started implementing that but never finished, had some issues. I can try to rebase that and push what I have, maybe you can figure out what's missing.

Yes, all methods must be switched to their corresponding `async_` version. But only once we have the async version of pytuya in place. So I will do that once I'm done.

rospogrigio commented:
> Reloading requires the `reload` service to work. I started implementing that but never finished, had some issues. I can try to rebase that and push what I have, maybe you can figure out what's missing.

OK I can try, I'd be happy to help. I initially thought it had to do with #30 but then I realized that it was only related to config flow.
 
> Yes, all methods must be switched to their corresponding `async_` version. But only once we have the async version of pytuya in place. So I will do that once I'm done.

OK, I just tried to simply rename the methods and I actually don't see any different behavior, but I thought it had to be related to pytuya refactoring also. 👍

postlund commented:
> OK I can try, I'd be happy to help. I initially thought it had to do with #30 but then I realized that it was only related to config flow.

That was mainly to support automatic updates when changing something via options.

> OK, I just tried to simply rename the methods and I actually don't see any different behavior, but I thought it had to be related to pytuya refactoring also. 👍

It will work, but the problem is that when you call something that is blocking from such a method, it will block the entire event loop. So **nothing** else will run when doing so, which is very very bad.

rospogrigio commented:
> 1. weeks ago I submitted a logo for localtuya custom integration as you suggested but it was never approved... is there somebody I should ask?

Looks like I made a PR in my own repo/fork instead of the official homeassistant one... 😅 🙄 Ok now I made the official PR, let's see if someone will review and approve it... 😉

rospogrigio commented on custom_components/localtuya/__init__.py:
```diff
@@ -20,9 +20,11 @@
-        stop_cmd: "stop" # Optional
+        open_close_cmds: ["on_off","open_close"] # Optional, default: "on_off"
+        positioning_mode: ["none","position","fake"] # Optional, default: "none"
+        currpos_dps: 3 # Optional, required only for "position" mode
```
Right, I've been thinking about this too actually. 👍

postlund commented on custom_components/localtuya/const.py:
```diff
@@ -14,9 +14,11 @@
-CONF_STOP_CMD = "stop_cmd"
+CONF_OPENCLOSE_CMDS = "open_close_cmds"
+CONF_POSITIONING_MODE = "positioning_mode"
+CONF_CURRENT_POSITION_DP = "currpos_dp"
```
Should be `"current_position_dp"` according to earlier. Same for next line.

postlund reviewed (changes requested):
One minor thing then we are done!

rospogrigio commented:
OK @postlund , all done, let's close this! I also have prepared the documentation 😉

postlund reviewed (approved):
Let's party and dance! 💃🎉
